### ISSUE TITLE
lzf_compress parameters differ from pecl php lzf_compress implementation

### ISSUE THREAD

**Issue by TheLevti** [phpredis/phpredis#1477]
### Expected behaviour

When using the php's `lzf_compress` function (when installing it via `pecl install lzf`), the result differs from the phpredis usage of the c `lzf_compress` function.

When for example the value is `abc abc abc`, using the php's `lzf_compress` function results in the following byte code (when viewing through redis-cli MONITOR) `\x04abc a@\x03\x01bc`.

### Actual behaviour

Enabling compression for this extension and connection and using the same value `abc abc abc`, the stored result is `abc abc abc`. It appears that with the specific parameters with which lzf_compress is called, differ from the ones used in such a way that when compression is not worth it, it will just return the original string. E.g. the string is not long enough. By using 4 times abc, it starts to work.

Why it is important that both libraries are using the same parameters is that sometimes I have to match up the value stored in redis. I know that the library used lzf for compression and I expect that if I use the same function to compress a string on php side, the result will be equal.

One use case is that this library does not serialize nor compress arguments of an EVAL call, it is up to the user to do that before sending the script to redis. By checking the currently set options (getOption()), I can determine if I have to manually compress the argument. But if the result is different from the phpredis usage, I will not be able to for example compare the stored KEY with the given argument in the EVAL script.

As a sidenote, you implemented a nice utility function ->_serialize() on the connection, so that no matter what I get the serialized data, the same way that the connection is configured. It would be really nice to have that for compression as well instead of checking doing something like the following to determine if I have to use lzf_compress myself:

```php
$lzfCompression = false;
if (\defined(Redis::class . '::COMPRESSION_LZF') &&
    Redis::COMPRESSION_LZF === $redis->getOption(Redis::OPT_COMPRESSION) &&
    \function_exists('\lzf_compress')
) {
    $lzfCompressed = \lzf_compress($data);
}
```

### I'm seeing this behaviour on
- OS: Ubuntu 18.04
- Redis: 5.0.2, 4.0.11
- PHP: 7.2.10
- phpredis: 4.2.0
- LZF: 1.6.6 (bundled with liblzf 3.6, same as phpredis is using)

### Steps to reproduce, backtrace or example script

I stumbled upon this, when I tried to use the https://github.com/php-lock/lock library, which uses a redlock implementation (https://redis.io/topics/distlock) to lock resources in a decentralized network. But, because I enabled compression, I discovered that this library forgot to in addition to call `$redis->_serialize($lockKey);` to also check if compression is enabled and if the EVAL arguments need to be compressed. When I tried to fix that, I discovered that the locking fails, because the stored KEY is different than against which I check (which was compressed by php's lzf_compress function).

The difference between the pecl lfz and phpredis lfz implementation is in the file `library.c:2136`

```c
PHP_REDIS_API int
redis_pack(RedisSock *redis_sock, zval *z, char **val, strlen_t *val_len TSRMLS_DC)
{
    char *buf;
    int valfree;
    strlen_t len;
#ifdef HAVE_REDIS_LZF
    char *data;
    uint32_t res;
#endif

    valfree = redis_serialize(redis_sock, z, &buf, &len TSRMLS_CC);
    switch (redis_sock->compression) {
        case REDIS_COMPRESSION_LZF:
#ifdef HAVE_REDIS_LZF
            data = emalloc(len);

            res = lzf_compress(buf, len, data, len - 1); // HERE THE LAST PARAMETER DIFFERS

            if (res > 0 && res < len) {
                if (valfree) efree(buf);
                *val = data;
                *val_len = res;
                 return 1;
            }
            efree(data);
#endif
            break;
    }
    *val = buf;
    *val_len = len;
    return valfree;
}
```

I saw the suggestion of the lzf source code is to use len - 1 for the last parameter (https://github.com/cmusatyalab/openisr/blob/master/crypto/lzf.c#L225), but the pecl php extension uses 
I think len + 128 (https://github.com/php/pecl-file_formats-lzf/blob/master/lzf.c#L141). Would it be possible to match their implementation? The current solution just saves a few bytes by not compressing too short strings if I understood it correctly. As I am no C programmer, can someone check both implementations and tell me what the difference is? It would really help to solve this problem, otherwise I can not use this extension together with compression.

Any help would be appreciated.

### I've checked
- [x] There is no similar issue from other users
- [x] Issue isn't fixed in `develop` branch

**Comment by yatsukhnenko** [comment]
@remicollet, sorry to bother you, could you explain why pecl-file_formats-lzf uses LZF_MARGIN?

**Comment by remicollet** [comment]
@yatsukhnenko sorry, no idea, I only took lead on this extension for PHP 7 compatibility.

**Comment by remicollet** [comment]
@yatsukhnenko after a quick look, it may happen that out buffer is a bit larger than in buffer. LZF_MARGIN is there to ensure compression always works.

From lzf header comment.
```
 * If the output buffer is not large enough or any error occurs return 0,
 * otherwise return the number of bytes used, which might be considerably
 * more than in_len (but less than 104% of the original size), so it
 * makes sense to always use out_len == in_len - 1), to ensure _some_
 * compression, and store the data uncompressed otherwise (with a flag, of
 * course.
```

In LZF extension case, out buffer is larger than in buffer to avoid failure

IMHO, having uncompressed data stored where  compressed data are expected can raise bad issue.

**Comment by yatsukhnenko** [comment]
> LZF_MARGIN is there to ensure compression always works

This makes sense when LZF_MARGIN is computed dynamically from in buffer size.

**Comment by remicollet** [comment]
I agress, LZF_MARGIN should be 4% of in buffer len (probably with a minimal value)

**Comment by yatsukhnenko** [comment]
@TheLevti right now I see two possible solutions
1. increase out buffer size to 104% of in buffer size
2. add an option to set margin to be compatible with file_formats-lzf extension

**Comment by yatsukhnenko** [comment]
@remicollet are you going to make any changes in file_formats-lzf extension?

**Comment by yatsukhnenko** [comment]
OK, thanks! I'll make changes in phpredis and than will create a PR with link to this issue :)

**Comment by TheLevti** [comment]
Just wow. Thank you guys. I never expected such a quick and collaborative response!

**Comment by yatsukhnenko** [comment]
@TheLevti this doesn't mean I'll do this change quickly :smile:

**Comment by TheLevti** [comment]
If you exactly tell me what needs to be done, I can try and help. But my C level is below average.

**Comment by TheLevti** [comment]
I will try today. Thanks.

**Comment by yatsukhnenko** [comment]
I guess, `make && make install` should be enough. Which version of extension do you see in `phpinfo`?

**Comment by TheLevti** [comment]
4.2.0 it says. i removed the one from pecl first, made sure that php fails if extension is missing, run `make install`, and then got this version.

**Comment by yatsukhnenko** [comment]
> `abc`, while lzf_compress returned `\tabc`

This means that comment `but less than 104% of the original size` is a lie

**Comment by remicollet** [comment]
> This means that comment but less than 104% of the original size is a lie

The reason why I propose to keep LZF margin ;)

**Comment by yatsukhnenko** [comment]
@jrchamp are you going to create PR?

**Comment by remicollet** [comment]
@jrchamp I like your way, avoiding float, BTW

> size = min(UINT_MAX, len + max(LZF_MARGIN, (uint)(len / 25)));

This will raise interger overflow. Seems better to:

	sizz =  i + min(UINT_MAX-i, max(LZF_MARGIN, (uint)(i / 25)));

**Comment by yatsukhnenko** [comment]
@TheLevti now compression should work as you expected.

**Comment by TheLevti** [comment]
Thanks guys. It will be a busy weekend. I will try to test it as soon as possible.

**Comment by TheLevti** [comment]
Ok, I just tested it and it works! Thanks a lot.

```
// This is set by phpredis.
1545426727.919770 [0 192.168.0.1:51042] "SET" "lock_abc" "\x02abc" "ex" "4" "nx"

// Last argument at the end of the EVAL command are compressed with the lzf extension.
1545426728.920514 [0 192.168.0.1:51042] "EVAL" "if redis.call(\"get\",KEYS[1]) == ARGV[1] then\n                return redis.call(\"del\",KEYS[1])\n            else\n                return 0\n            end\n        " "1" "lock_abc" "\x02abc"

1545426728.920584 [0 lua] "get" "lock_abc"
1545426728.920600 [0 lua] "del" "lock_abc"
```

So lock could successfully be released.

**Comment by yatsukhnenko** [comment]
Fixed in `develop` and `4.3.0RC1` on PECL site
